src/rim_experiments/util/score_array.py

```diff
@@ -176,3 +202,84 @@ def __mul__(self, other):
         else:
             warnings.warn(f"falling back to dense to multiply with {other.__class__}")
             return self.eval() * other
+
+
+@dataclasses.dataclass(repr=False)
+class CustomLowRankDataFrame(ScoreExpression):
+    """ mimics a pandas dataframe with exponentiated low-rank structures
+    """
+    ind_logits: List[list]
+    col_logits: List[list]
+    sign: float
+    index: list
+    columns: list
+    activation: str = 'exponential'
+
+    def __post_init__(self):
+        assert self.ind_logits.shape[1] == self.col_logits.shape[1], "check hidden"
+        assert self.ind_logits.shape[0] == len(self.index), "check index"
+        assert self.col_logits.shape[0] == len(self.columns), "check columns"
+        assert self.activation in ['exponential','sigmoid','raw']
+
+
+    @property
+    def values(self):
+        if self.activation=='exponential':
+            return ExponentiatedLowRankValues(self.ind_logits, self.col_logits, self.sign)
+        elif self.activation=='sigmoid':
+            return SigmoidLowRankValues(self.ind_logits, self.col_logits, self.sign)
+        elif self.activation=='raw':
+            return DotProdLowRankValues(self.ind_logits, self.col_logits, self.sign)
+
+
+    def eval(self):
+        return pd.DataFrame(self.values.eval(), index=self.index, columns=self.columns)
+
+
+    def has_nan(self):
+        return np.isnan(self.ind_logits).any() or np.isnan(self.col_logits).any()
+
+
+    @property
+    def shape(self):
+        return (len(self.index), len(self.columns))
+
+
+    @property
+    def T(self):
+        return self.__class__(self.col_logits, self.ind_logits, self.sign,
+            self.columns, self.index, self.activation)
+
+
+    def reindex(self, index, axis=0, fill_value=float("nan")):
+        if axis==1:
+            return self.T.reindex(index, fill_value=fill_value).T
+
+        ind_logits = np.pad(self.ind_logits, ((0,1), (0,1)), constant_values=0)
+        #TODO: change fillna according to activation type
+        with np.errstate(divide='ignore'): # 0 -> -inf
+            ind_logits[-1, -1] = np.log(fill_value)
```

Comment:
`np.log(fill_value) if self.activation in ['exponential', 'sigmoid'] else fill_value # raw(0)`